src/tools/wasm-fuzz-lattices.cpp

```diff
@@ -0,0 +1,479 @@
+
+#include <optional>
+#include <random>
+#include <string>
+
+#include "analysis/lattice.h"
+#include "analysis/liveness-transfer-function.h"
+#include "analysis/reaching-definitions-transfer-function.h"
+
+#include "support/command-line.h"
+#include "tools/fuzzing.h"
+#include "tools/fuzzing/random.h"
+
+namespace wasm {
+using RandEngine = std::mt19937_64;
+using namespace analysis;
+
+// Helps printing error messages.
+std::string LatticeComparisonNames[4] = {
+  "No Relation", "Equal", "Less", "Greater"};
+std::string LatticeComparisonSymbols[4] = {"?", "=", "<", ">"};
+
+uint64_t getSeed() {
+  // Return a (truly) random 64-bit value.
+  std::random_device rand;
+  return std::uniform_int_distribution<uint64_t>{}(rand);
+}
+
+// Utility class which provides methods to check properties of the transfer
+// function and lattice of an analysis.
+template<typename Lattice, typename TransferFunction> struct AnalysisChecker {
+  Lattice& lattice;
+  TransferFunction& transferFunction;
+  std::string latticeName;
+  std::string transferFunctionName;
+  uint64_t latticeElementSeed;
+  Name funcName;
+
+  AnalysisChecker(Lattice& lattice,
+                  TransferFunction& transferFunction,
+                  std::string latticeName,
+                  std::string transferFunctionName,
+                  uint64_t latticeElementSeed,
+                  Name funcName)
+    : lattice(lattice), transferFunction(transferFunction),
+      latticeName(latticeName), transferFunctionName(transferFunctionName),
+      latticeElementSeed(latticeElementSeed), funcName(funcName) {}
+
+  void printFailureInfo(std::ostream& os) {
+    os << "Error for " << transferFunctionName << " and " << latticeName
+       << " at lattice element seed " << latticeElementSeed << " and function "
+       << funcName << ".\n";
+  }
+
+  // Checks reflexivity of a lattice element, i.e. x = x.
+  void checkReflexivity(typename Lattice::Element& element) {
+    LatticeComparison result = lattice.compare(element, element);
+    if (result != LatticeComparison::EQUAL) {
+      std::stringstream ss;
+      printFailureInfo(ss);
+      ss << "Element ";
+      element.print(ss);
+      ss << " is not reflexive.\n";
+      Fatal() << ss.str();
+    }
+  }
+
+  // Anti-Symmetry is defined as x <= y and y <= x imply x = y. Due to the
+  // fact that the compare(x, y) function of the lattice explicitly tells
+  // us if two lattice elements are <, =, or = instead of providing a
+  // <= comparison, it is not useful to check for anti-symmetry as it is defined
+  // in the fuzzer.
+
```

Comment:
```suggestion
  //
```
(This is all one big comment.)